cpp/src/neighbors/ivf_pq/ivf_pq_build_common.cu

```diff
@@ -311,6 +311,27 @@ void extract_centers(raft::resources const& res,
                                   stream));
 }
 
+void extract_centers(raft::resources const& res,
```

Comment:
Since we need both -- the host and device mdspan functions for `extract_centers`, we can have a pointer based function in the detail namespace, rather than duplicating the function for host mdspan. Then the host and device mdspan functions in the src file can call that pointer based function.

Reply:
I don't think we are going to be able to avoid this distinction here, and it's actually very low impact here to have the branched logic (in general, we want to move away from pointer-based functions across the codebase, not just in the public APIs... but we started w/ the public APIs). Not all device functions are super expensive.

Reply:
I am concerned that the content of the function is an exact copy except the function signature. In some other places we have had public APIs that are agnostic of what mdspan is being used (so the function argument is`raft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, data_accessor>`)

Reply:
I changed in the last commit to reduce the cut and paste code here (not using a pointer based api - instead just using a template version on the output mdspan type).  lmk what you all think

Reply:
On the other hand, if we are okay with the two functions because it is not too expensive, that is also fine :))

Reply:
> I am concerned that the content of the function is an exact copy except the function signature. In some other places we have had public APIs that are agnostic of what mdspan is being used (so the function argument israft::mdspan<const T, raft::matrix_extent<int64_t>, raft::row_major, data_accessor>)

I'm in favor of accepting general `mdspan`. Not in favor of accepting raw pointers. Raw pointers are prone to too much error- it would be a giant step backwards for our code.

Reply:
this commit here is the one I meant https://github.com/rapidsai/cuvs/pull/888/commits/7080e2628b61d8c6fe6e2e4621b00ef2cad4f292 (takes a general mdspan )

Reply:
Looks good. Just one nit pick (feel free to ignore since it is not that important). We can move detail namespace stuff to `ivf_pq_build.cuh`

Reply:
> Looks good. Just one nit pick (feel free to ignore since it is not that important). We can move detail namespace stuff to `ivf_pq_build.cuh`

moved in ed474d1